File: momentliving-ai-service/src/main/java/com/momentliving/tools/ShopTools.java
package com.momentliving.tools;

import com.momentliving.api.client.ShopClient;
import com.momentliving.context.UserHolder;
import com.momentliving.dto.ShopQueryDTO;
import com.momentliving.result.Result;
import com.momentliving.vo.ReviewVO;
import com.momentliving.vo.ShopVO;
import com.momentliving.vo.UserVO;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.model.ToolContext;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;
import java.util.Map;

/**
 * 商铺查询工具（Function Calling）：数据一律来自 shop-service 实时接口，AI 禁止编造。
 * 工具方法只读公开数据，不依赖用户身份（公共接口）。
 */
@Slf4j
@Component
public class ShopTools {

    @Resource
    private ShopClient shopClient;

    @Tool(description = "根据关键词搜索商铺，返回商铺列表（名称、区域、地址、评分、人均）")
    public List<Map<String, Object>> searchShops(
            @ToolParam(description = "搜索关键词，如火锅店、咖啡") String keyword,
            @ToolParam(description = "排序方式：score评分最高 / sold销量最高 / other默认", required = false) String sort) {
        long start = System.currentTimeMillis();
        ShopQueryDTO query = new ShopQueryDTO();
        query.setName(keyword);
        query.setCurrent(1);
        Result<List<ShopVO>> result = shopClient.searchShops(query);
        List<ShopVO> shops = result.getData() == null ? List.of() : result.getData();
        if ("score".equals(sort)) {
            shops = shops.stream()
                    .sorted(Comparator.comparing(ShopVO::getScore, Comparator.nullsLast(Comparator.reverseOrder())))
                    .toList();
        } else if ("sold".equals(sort)) {
            shops = shops.stream()
                    .sorted(Comparator.comparing(ShopVO::getSold, Comparator.nullsLast(Comparator.reverseOrder())))
                    .toList();
        }
        List<Map<String, Object>> data = shops.stream().limit(10).map(this::toMap).toList();
        logTool("searchShops", Map.of("keyword", keyword), data.size(), start);
        return data;
    }

    @Tool(description = "根据商铺ID获取商铺详情（名称、地址、人均、评分、营业时间）")
    public Map<String, Object> getShopDetail(@ToolParam(description = "商铺ID") Long shopId) {
        long start = System.currentTimeMillis();
        Result<ShopVO> result = shopClient.getShop(shopId);
        Map<String, Object> data = result.getData() == null
                ? Map.of("error", "商铺不存在")
                : toMap(result.getData());
        logTool("getShopDetail", Map.of("shopId", shopId), data.size(), start);
        return data;
    }

    @Tool(description = "获取商铺的用户评价列表（内容、星级、昵称）")
    public List<Map<String, Object>> getShopReviews(@ToolParam(description = "商铺ID") Long shopId,
                                                    ToolContext toolContext) {
        long start = System.currentTimeMillis();
        // ⚠️ /review/** 在 shop-service 本地也要求登录（只有 /shop/** 公开），
        // 流式场景工具跑在无身份线程上，必须临时补 X-User-Id，否则 Feign 会被 401
        Long userId = resolveUserId(toolContext);
        if (userId == null) {
            log.warn("AI工具调用 getShopReviews 缺少用户身份，拒绝查询");
            return List.of(Map.of("error", "未获取到用户身份，请重新登录后再试"));
        }
        boolean tempSaved = false;
        if (UserHolder.getUser() == null) {
            UserHolder.saveUser(UserVO.builder().id(userId).build());
            tempSaved = true;
        }
        List<Map<String, Object>> data;
        try {
            Result<List<ReviewVO>> result = shopClient.getShopReviews(shopId, 1);
            data = (result.getData() == null ? List.<ReviewVO>of() : result.getData())
                    .stream()
                    .limit(10)
                    .map(r -> Map.<String, Object>of(
                            "rating", r.getRating() == null ? 5 : r.getRating(),
                            "content", abbreviate(r.getContent(), 100),
                            "nickName", r.getNickName() == null ? "匿名用户" : r.getNickName()))
                    .toList();
        } finally {
            if (tempSaved) {
                UserHolder.removeUser();
            }
        }
        logTool("getShopReviews", Map.of("shopId", shopId), data.size(), start);
        return data;
    }

    /** 身份还原：优先 ToolContext（流式场景），回退 UserHolder（同步场景） */
    private Long resolveUserId(ToolContext toolContext) {
        if (toolContext != null && toolContext.getContext().get("aiUserId") != null) {
            return Long.valueOf(toolContext.getContext().get("aiUserId").toString());
        }
        return UserHolder.getUser() != null ? UserHolder.getUser().getId() : null;
    }

    private Map<String, Object> toMap(ShopVO shop) {
        return Map.of(
                "shopId", shop.getId(),
                "name", shop.getName() == null ? "" : shop.getName(),
                "area", shop.getArea() == null ? "" : shop.getArea(),
                "address", shop.getAddress() == null ? "" : shop.getAddress(),
                "avgPrice", shop.getAvgPrice() == null ? 0 : shop.getAvgPrice(),
                "score", shop.getScore() == null ? 0 : shop.getScore(),
                "sold", shop.getSold() == null ? 0 : shop.getSold(),
                "openHours", shop.getOpenHours() == null ? "" : shop.getOpenHours());
    }

    private String abbreviate(String s, int max) {
        if (s == null) return "";
        return s.length() <= max ? s : s.substring(0, max) + "...";
    }

    private void logTool(String name, Object params, int resultSize, long start) {
        log.info("AI工具调用 name={}, params={}, resultSize={}, costMs={}",
                name, params, resultSize, System.currentTimeMillis() - start);
    }
}

File: momentliving-ai-py/app/tools/shop_tools.py
"""商铺工具（替代 Spring ShopTools，含身份敏感调用）。"""
import httpx
from langchain_core.tools import tool
from pydantic import Field

from app.clients.auth_header import get_identity_headers
from app.config import get_settings


def _abbreviate(s: str | None, max_len: int) -> str:
    if not s:
        return ""
    return s if len(s) <= max_len else s[:max_len] + "..."


def _shop_to_map(shop: dict) -> dict:
    return {
        "shopId": shop.get("id"),
        "name": shop.get("name") or "",
        "area": shop.get("area") or "",
        "address": shop.get("address") or "",
        "avgPrice": shop.get("avgPrice") or 0,
        "score": shop.get("score") or 0,
        "sold": shop.get("sold") or 0,
        "openHours": shop.get("openHours") or "",
    }


@tool
async def search_shops(
    keyword: str = Field(description="搜索关键词，如火锅店、咖啡"),
    sort: str | None = Field(default=None, description="排序方式：score评分最高 / sold销量最高 / other默认"),
) -> list[dict]:
    """
    根据关键词搜索商铺，返回商铺列表（名称、区域、地址、评分、人均、销量），
    回答"找店/推荐店铺"类问题必用

    Args:
        keyword: 搜索关键词
        sort: 排序方式，可选
    """
    settings = get_settings()

    async with httpx.AsyncClient(timeout=settings.upstream.timeout) as client:
        # 真实路径：ShopClient.searchShops(SpringQueryMap) 对应 GET /shop/of/name
        resp = await client.get(
            f"{settings.upstream.shop_service_url}/shop/of/name",
            params={"name": keyword, "current": 1, "sort": sort or "other"},
        )
        resp.raise_for_status()
        result = resp.json()

    shops = result.get("data") or []
    if sort == "score":
        shops = sorted(shops, key=lambda s: s.get("score") or 0, reverse=True)
    elif sort == "sold":
        shops = sorted(shops, key=lambda s: s.get("sold") or 0, reverse=True)

    return [_shop_to_map(s) for s in shops[:10]]


@tool
async def get_shop_detail(shop_id: int = Field(description="商铺ID")) -> dict:
    """根据商铺ID获取商铺详情（名称、区域、地址、人均、评分、销量、营业时间）"""
    settings = get_settings()

    async with httpx.AsyncClient(timeout=settings.upstream.timeout) as client:
        resp = await client.get(f"{settings.upstream.shop_service_url}/shop/{shop_id}")
        resp.raise_for_status()
        result = resp.json()

    shop = result.get("data")
    if not shop:
        return {"error": "商铺不存在"}
    return _shop_to_map(shop)


@tool
async def get_shop_reviews(shop_id: int = Field(description="商铺ID")) -> list[dict]:
    """
    获取商铺的用户评价列表（星级、内容、昵称），回答"这家店评价怎么样"类问题可用

    Args:
        shop_id: 商铺ID
    """
    settings = get_settings()
    headers = get_identity_headers()

    # 上游 shop-service 的 /review/** 路径要求登录态（X-User-Id）
    if "X-User-Id" not in headers and "X-Merchant-Id" not in headers:
        return [{"error": "未获取到用户身份，请重新登录后再试"}]

    async with httpx.AsyncClient(timeout=settings.upstream.timeout) as client:
        # 真实路径：ShopClient.getReviews(shopId, current) 对应 GET /review/list/{shopId}
        resp = await client.get(
            f"{settings.upstream.shop_service_url}/review/list/{shop_id}",
            params={"current": 1},
            headers=headers,
        )
        resp.raise_for_status()
        result = resp.json()

    reviews = result.get("data") or []
    return [
        {
            "rating": r.get("rating") or 5,
            "content": _abbreviate(r.get("content"), 100),
            "nickName": r.get("nickName") or "匿名用户",
        }
        for r in reviews[:10]
    ]
